### DISCUSSION TITLE
What is the best way to optimize bigger nuxt3 app?

### DISCUSSION THREAD

**Discussion by kstraszewski** [nuxt/nuxt#16888] [Questions]
Hi, I have a question what are the best practices for maintaining a small entry.js file? I manage to reduce significantly entry.css file, by using purge.css. But I have no idea to how I can reduce entry.js

**Comment by TheAlexLichter** [comment]
You can always apply code splitting (by loading components lazily via `<LazyXYZ>`, see https://v3.nuxtjs.org/guide/directory-structure/components#dynamic-imports) so your entry is smaller. But that's only advisable for components below the fold of pages.

Besides that, load libraries where needed and not globally.

**Reply by konradsienkowski** [reply]
How about libraries in composables? Since Nuxt automatically imports them

**Reply by TheAlexLichter** [reply]
Like VueUse you mean? That's only imported when you use it ☺️

**Reply by konradsienkowski** [reply]
No, I meant custom composables that use another library. E.g. `composables/useLibrary.ts`, where you import the library's ES module at the beginning of the file. Since `composables/` directory is included in [auto imports](https://v3.nuxtjs.org/guide/concepts/auto-imports), the library is going to be included in an entry file

**Reply by kstraszewski** [reply]
is there any way to lazy load composables?

**Reply by kstraszewski** [reply]
> You can always apply code splitting (by loading components lazily via `<LazyXYZ>`, see https://v3.nuxtjs.org/guide/directory-structure/components#dynamic-imports) so your entry is smaller. But that's only advisable for components below the fold of pages.
> 
> Besides that, load libraries where needed and not globally.

Shouldn't <LazyXYZ> be the same as defineAsyncComponent https://vuejs.org/guide/components/async.html#basic-usage?

**Comment by kstraszewski** [comment]
Ok I've manage to reduce entry js by 20%. But it is still not enough. 
Im adding this to nuxt.confit.ts -> vite
```
    build: {
      minify: 'terser',
      rollupOptions: {
        treeshake: {
          preset: 'smallest',
          annotations: true,
          moduleSideEffects: true,
          tryCatchDeoptimization: true,
          correctVarValueBeforeDeclaration: true,
        },
      },
    },
```

I've realized that my entry.js have lots of services that should be dynamic imported. Moving them to composable adds them to entry.js unfortunately. And also I have no idea when nuxt adds service to entry and when it load it dynamically. I have 2 services in service folder. 1 is dynamically imported and the otherone (the bigger one) is included to entry.js. Both are used a lot in pages and also in components and other services.
